### ISSUE TITLE
Would compatibility with `podman` / `podman-compose` be welcome?

### ISSUE THREAD

**Issue by meatballhat** [pypi/warehouse#15981]
* **Date @ Event:** 5/20/2023 @ PyCon US Sprints
* **Scenario:** Running `make build`
* **OS environment:** Arch (rolling, circa now-ish)
* **Things you used:** warehouse
* **Python version and origin:** N/A (?)

## Please highlight your log using this legend:
- [x] :heart_eyes: Yay, this is awesome
- [ ] :frowning: This is a little annoying, I'm frustrated
- [ ] :rage: I'd quit if I weren't specifically trying to improve things.

## What were you trying to do, and what didn't work?
Running `make build` resulted in:

```sh-session
make[1]: Entering directory '/home/me/code/pypi/warehouse'
# Build our base container for this project.
docker compose build --build-arg IPYTHON=no --force-rm base
>>>> Executing external compose provider "/home/me/.local/bin/docker-compose". Please refer to the documentation for details. <<<<

usage: docker-compose [-h] [-v] [--in-pod in_pod] [--pod-args pod_args] [--env-file env_file] [-f file] [--profile profile] [-p PROJECT_NAME] [--podman-path PODMAN_PATH] [--podman-args args] [--podman-pull-args args]
                      [--podman-push-args args] [--podman-build-args args] [--podman-inspect-args args] [--podman-run-args args] [--podman-start-args args] [--podman-stop-args args] [--podman-rm-args args] [--podman-volume-args args]
                      [--no-ansi] [--no-cleanup] [--dry-run] [--parallel PARALLEL] [--verbose]
                      {help,version,wait,systemd,pull,push,build,up,down,ps,run,exec,start,stop,restart,logs,config,port,pause,unpause,kill,stats} ...
docker-compose: error: unrecognized arguments: --force-rm
Error: executing /home/me/.local/bin/docker-compose build --build-arg IPYTHON=no --force-rm base: exit status 2
make[1]: *** [Makefile:22: .state/docker-build-base] Error 2
make[1]: Leaving directory '/home/me/code/pypi/warehouse'
make: *** [Makefile:53: build] Error 2
```

I have `podman` and `podman-compose` locally symlinked as `docker` and `docker-compose`, respectively. The usage of `docker-compose` within Warehouse doesn't play nicely with this, which isn't much of a surprise, but I'd like to know if introducing Podman compatibility would be welcome.

## Ideally, what should have happened instead?
`make build` succeeds!

**Comment by miketheman** [comment]
Hey @meatballhat !

Thanks for the detailed issue.

Having not used podman much, do you know how much might change as a result?

We currently support Docker and docker-like contexts, such as OrbStack which acts as a drop in replacement.

If podman doesn't offer a similar interface, I think it'd come down to how much would have to change to afford that compatibility.

Another risk to consider is that I don't think any of the current maintainers use podman, so on the event of failure/friction we may not be able to support users, so please also consider how we might exercise podman in an automatic testing approach (we have a GitHub action today that runs through some of the setup steps once a day).

**Comment by meatballhat** [comment]
@miketheman Thank you! Yep, very much understood. My general experience with using podman as drop-in like you describe with OrbStack has been pretty smooth, but of course there are moments like this. If you're OK with me at least taking a look, I'm happy to see just how much change is needed and get a patch ready for review. If it ends up looking like too much of a stretch, I'm very much willing to follow the happy path with Real Docker 😁 WDYT?

**Comment by miketheman** [comment]
If you want to see what's involved, I definitely do not want to discourage that.

**Comment by meatballhat** [comment]
tl;dr reporting back for future explorers: This will take _enough doing_ that I think the time would be better spent [contributing to `podman-compose`](https://github.com/containers/podman-compose/blob/cac90f69b816fd8e5c65cfd05491f91ee1404f2f/CONTRIBUTING.md#missing-commands-help-needed).

The longer version is that the high-level `make` targets will not work with `podman` and `podman compose` / `podman-compose` in their current implementations. I started to parameterize some of the incompatible flags, but it makes the `make` targets a good bit more cluttered and difficult to follow. In the end, I felt like I was introducing too much branching and conditional weirdness. This is all very IMHO!

**Comment by miketheman** [comment]
Thanks @meatballhat for digging into it. Happy to review another patch if this comes up again once podman supports more of the same/similar interface commands.
